anndata/_core/anndata.py

```diff
@@ -627,9 +627,7 @@ def X(self) -> Optional[Union[np.ndarray, sparse.spmatrix, ArrayView]]:
     def X(self, value: Optional[Union[np.ndarray, sparse.spmatrix]]):
         if value is None:
             if self.is_view:
-                raise ValueError(
-                    "Copy the view before setting the data matrix to `None`."
-                )
+                self._init_as_actual(self.copy())
```

Comment:
I think this should error if the view is of a backed object, since I'm not sure we have the logic to deal with that at the moment.

Reply:
I think these checks need to be flipped, so the lower error get's thrown, not an error from trying to copy.

At the moment:

```python
from anndata.tests.helpers import gen_adata
import anndata as ad

gen_adata((10, 20)).write_h5ad("tmp.h5ad")
b = ad.read_h5ad("tmp.h5ad", backed="r")
v = b[::2, ::2]
del v.X
```

```pytb
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-11-6b92c05b346f> in <module>
----> 1 del v.X

~/github/anndata/anndata/_core/anndata.py in X(self)
    685     @X.deleter
    686     def X(self):
--> 687         self.X = None
    688 
    689     @property

~/github/anndata/anndata/_core/anndata.py in X(self, value)
    628         if value is None:
    629             if self.is_view:
--> 630                 self._init_as_actual(self.copy())
    631             if self.isbacked:
    632                 raise ValueError("Not implemented.")

~/github/anndata/anndata/_core/anndata.py in copy(self, filename)
   1460 
   1461             if filename is None:
-> 1462                 raise ValueError(
   1463                     "To copy an AnnData object in backed mode, "
   1464                     "pass a filename: `.copy(filename='myfilename.h5ad')`."

ValueError: To copy an AnnData object in backed mode, pass a filename: `.copy(filename='myfilename.h5ad')`.
```

Also the error should be more descriptive, and probably be an actual `NotImplementedError`.

Maybe:

```python
raise NotImplementedError("Cannot currently remove data matrix from backed object.")
```

Reply:
Done! :)